### PR TITLE
Fix bug with TrainingValidationSplit and custom index

### LINKED ISSUE #1126: IndexError: positional indexers are out-of-bounds
### Reproducer

Note: the example below works when `n_rows` is small

```
import evalml
import pandas as pd
automl = evalml.AutoMLSearch(problem_type="binary")

n_rows = 1000000 
X = pd.DataFrame({"f_1": [1, 2, 3]*n_rows})
X.index = pd.Series(range(X.shape[0])) + 1
y = pd.Series([True, False, True]*n_rows)

automl.search(X, y, data_checks='disabled')
```

this give the follow stacktrace

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
~/.pyenv/versions/3.6.9/envs/tmp-install/lib/python3.6/site-packages/pandas/core/indexing.py in _get_list_axis(self, key, axis)
   2110         try:
-> 2111             return self.obj._take_with_is_copy(key, axis=axis)
   2112         except IndexError:

~/.pyenv/versions/3.6.9/envs/tmp-install/lib/python3.6/site-packages/pandas/core/generic.py in _take_with_is_copy(self, indices, axis, **kwargs)
   3408         """
-> 3409         result = self.take(indices=indices, axis=axis, **kwargs)
   3410         # Maybe set copy if we didn't actually change the index.

~/.pyenv/versions/3.6.9/envs/tmp-install/lib/python3.6/site-packages/pandas/core/generic.py in take(self, indices, axis, is_copy, **kwargs)
   3394         new_data = self._data.take(
-> 3395             indices, axis=self._get_block_manager_axis(axis), verify=True
   3396         )

~/.pyenv/versions/3.6.9/envs/tmp-install/lib/python3.6/site-packages/pandas/core/internals/managers.py in take(self, indexer, axis, verify, convert)
   1385         if convert:
-> 1386             indexer = maybe_convert_indices(indexer, n)
   1387 

~/.pyenv/versions/3.6.9/envs/tmp-install/lib/python3.6/site-packages/pandas/core/indexers.py in maybe_convert_indices(indices, n)
    212     if mask.any():
--> 213         raise IndexError("indices are out-of-bounds")
    214     return indices

IndexError: indices are out-of-bounds

During handling of the above exception, another exception occurred:

IndexError                                Traceback (most recent call last)
<ipython-input-24-2d762c4df3d4> in <module>
      8 y = pd.Series([True, False, True]*n_rows)
      9 
---> 10 automl.search(X, y, data_checks='disabled')

~/.pyenv/versions/3.6.9/envs/tmp-install/lib/python3.6/site-packages/evalml/automl/automl_search.py in search(self, X, y, data_checks, feature_types, show_iteration_plot)
    410         self._start = time.time()
    411 
--> 412         should_terminate = self._add_baseline_pipelines(X, y)
    413         if should_terminate:
    414             return

~/.pyenv/versions/3.6.9/envs/tmp-install/lib/python3.6/site-packages/evalml/automl/automl_search.py in _add_baseline_pipelines(self, X, y)
    543                                 self._start)
    544 
--> 545                 baseline_results = self._compute_cv_scores(baseline, X, y)
    546                 self._baseline_cv_score = baseline_results["cv_score_mean"]
    547                 self._add_result(trained_pipeline=baseline,

~/.pyenv/versions/3.6.9/envs/tmp-install/lib/python3.6/site-packages/evalml/automl/automl_search.py in _compute_cv_scores(self, pipeline, X, y)
    563         for i, (train, test) in enumerate(self.data_split.split(X, y)):
    564             logger.debug(f"\t\tTraining and scoring on fold {i}")
--> 565             X_train, X_test = X.iloc[train], X.iloc[test]
    566             y_train, y_test = y.iloc[train], y.iloc[test]
    567             objectives_to_score = [self.objective] + self.additional_objectives

~/.pyenv/versions/3.6.9/envs/tmp-install/lib/python3.6/site-packages/pandas/core/indexing.py in __getitem__(self, key)
   1766 
   1767             maybe_callable = com.apply_if_callable(key, self.obj)
-> 1768             return self._getitem_axis(maybe_callable, axis=axis)
   1769 
   1770     def _is_scalar_access(self, key: Tuple):

~/.pyenv/versions/3.6.9/envs/tmp-install/lib/python3.6/site-packages/pandas/core/indexing.py in _getitem_axis(self, key, axis)
   2127         # a list of integers
   2128         elif is_list_like_indexer(key):
-> 2129             return self._get_list_axis(key, axis=axis)
   2130 
   2131         # a single integer

~/.pyenv/versions/3.6.9/envs/tmp-install/lib/python3.6/site-packages/pandas/core/indexing.py in _get_list_axis(self, key, axis)
   2112         except IndexError:
   2113             # re-raise with different error message
-> 2114             raise IndexError("positional indexers are out-of-bounds")
   2115 
   2116     def _getitem_axis(self, key, axis: int):

IndexError: positional indexers are out-of-bounds
```

### workaround

This works

```
automl.search(X.reset_index(), y, data_checks='disabled')
```

### Notes

This seems to only occur once the data is above a certain number of rows, so i suspect it has to do with the different data splitter we use in that case

### PR DESCRIPTION
**Pull request by freddyaboulton**
### Pull Request Description
Fixes #1126

-----
*After creating the pull request: in order to pass the **release_notes_updated** check you will need to update the "Future Release" section of* `docs/source/release_notes.rst` *to include this pull request by adding :pr:`123`.*

### CODE REVIEW

**Review by jeremyliweishih — APPROVED** [review]
Looks good to me. Are there any side effects of using an numpy array instead of pandas?

**Comment by freddyaboulton** [comment]
@jeremyliweishih Not that I'm aware of! `iloc` works with pandas, numpy, lists and I believe the sklearn splitters return numpy arrays anyways.

**Review by bchen1116 — APPROVED** [review]
LGTM!

File: evalml/tests/automl_tests/test_training_validation_split.py [hunk 1]
```diff
@@ -51,5 +52,15 @@ def test_tvsplit_stratify():
...
+
+
+@pytest.mark.parametrize("random_state", [0, 11, 57, 99, 1000])
+def test_tvsplit_always_within_bounds_with_custom_index(random_state):
+    N = 11000
+    X = pd.DataFrame({'col1': np.arange(0, N)}, index=np.arange(20000, 20000 + N))
+    splitter = TrainingValidationSplit(train_size=0.75, shuffle=True, random_state=random_state)
+    splits = splitter.split(X, y=None)
+    assert np.all(np.logical_and(splits[0][0] < N, splits[0][0] >= 0))
+    assert np.all(np.logical_and(splits[0][1] < N, splits[0][1] >= 0))
```
Comment (dsherry) [inline evalml/tests/automl_tests/test_training_validation_split.py:66]: @freddyaboulton I see you defined a custom index here. That makes sense. Shouldn't we check that the output still has the custom indices?
```python
np.testing.assert_equal(20000 + np.arange(0, round(int(0.75*N))), splits[0][0].index)
np.testing.assert_equal(20000 + np.arange(round(int(0.75*N)), N), splits[0][1].index)
```

Comment (freddyaboulton) [reply evalml/tests/automl_tests/test_training_validation_split.py:66]: Our splitters don't split the data, they just provide the indices at which the data should be split. But after we call .iloc on these indices, the custom index will be preserved. Should I add that to the test?

File: evalml/automl/data_splitters/training_validation_split.py [hunk 2]
```diff
@@ -41,5 +42,5 @@ def split(self, X, y=None):
             Returns:
                 list: indices to split data into training and test set
         """
-        train, test = train_test_split(X.index, test_size=self.test_size, train_size=self.train_size, shuffle=self.shuffle, stratify=self.stratify, random_state=self.random_state)
+        train, test = train_test_split(np.arange(X.shape[0]), test_size=self.test_size, train_size=self.train_size, shuffle=self.shuffle, stratify=self.stratify, random_state=self.random_state)
```
Comment (dsherry) [inline evalml/automl/data_splitters/training_validation_split.py:45]: Ok, yep, I think this makes sense and aligns with sklearn's behavior.

The following
```python
from sklearn.model_selection import KFold
N = 11000
X = pd.DataFrame({'col1': np.arange(0, N)}, index=np.arange(20000, 20000 + N))
sk_splits = list(KFold(2, random_state=0).split(X))
sk_splits[0]
```
outputs splits generated in terms of the row number, ignoring the custom index:
```python
(array([ 5500,  5501,  5502, ..., 10997, 10998, 10999]),
 array([   0,    1,    2, ..., 5497, 5498, 5499]))
```
and therefore [our automl search code which uses `iloc`](https://github.com/alteryx/evalml/blob/main/evalml/automl/automl_search.py#L638) to make the dataframes for each split will work correctly.

For the TrainingValidationSplit, using `np.arange(X.shape[0])` as input to `train_test_split` as you've done here produces the same behavior: the result is the row number, and completely ignores the custom index, which is appropriate for `iloc`.

Great fix!

**Review by dsherry — APPROVED** [review]
@freddyaboulton nice! Thanks for chasing this down. I left one testing request

File: evalml/tests/automl_tests/test_training_validation_split.py (hunk 1 above)
Comment (dsherry) [reply evalml/tests/automl_tests/test_training_validation_split.py:66]: Ah, you're right, never mind!

Comment (dsherry) [reply evalml/tests/automl_tests/test_training_validation_split.py:66]: If I were gonna add another unit test concerning this stuff, it'd be an automl test which runs with `max_pipelines=2` and a dataset with a custom index, mocks pipeline `fit` and makes sure that the data handed to the pipelines is exactly what we expect it to be. Not required for this PR though.

**Pull request merged**